File: carbon-jdk-code/libs/carbonlog-model/src/main/java/ch/burguiere/carbonlog/model/CarbonLog.java
package ch.burguiere.carbonlog.model;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Stream;

public record CarbonLog(List<CarbonMeasurement> carbonMeasurements) {

    Double getRangeCarbonKgs(final Instant from, final Instant to, final Boolean inclusive) {

        final Predicate<CarbonMeasurement> filter;
        if (!inclusive) {
            filter = (cm) -> cm.dt().isAfter(from) && cm.dt().isBefore(to);
        } else {
            filter = (cm) -> (cm.dt().equals(from) || cm.dt().isAfter(from)) && (cm.dt().isBefore(to) || cm.dt().equals(to));
        }

        final var filtered = carbonMeasurements.stream()
                .filter(filter)
                .toList();
        return sumCO2kgs(filtered);
    }

    public Double getRangeCarbonKgs(final Instant from, final Instant to) {
        return getRangeCarbonKgs(from, to, false);
    }

    public CarbonLog copyAdding(final CarbonMeasurement measurement) {
        return copyAddingMultiple(List.of(measurement));
    }

    public CarbonLog copyAddingMultiple(final List<CarbonMeasurement> measurements) {
        return new CarbonLog(Stream.concat(this.carbonMeasurements.stream(), measurements.stream()).toList());
    }

    private static Double sumCO2kgs(final List<CarbonMeasurement> measurements) {
        if (measurements.isEmpty()) {
            return 0.0;
        }
        return measurements.stream()
                .map(CarbonMeasurement::co2Kg)
                .reduce(0.0, Double::sum);
    }

    public Double getCurrentYearCarbonKgs() {
        return sumCO2kgs(
                carbonMeasurements.stream()
                        .filter(cm -> cm.getYear().equals(LocalDate.now().getYear()))
                        .toList()
        );
    }

    public Double getTotalCarbonKgs() {
        return sumCO2kgs(carbonMeasurements);
    }
}

File: carbon-jdk-code/libs/carbonlog-model/src/jvmMain/kotlin/ch/burguiere/carbonlog/model/jvm/CarbonLog.kt
package ch.burguiere.carbonlog.model.jvm

import ch.burguiere.carbonlog.model.CoreCarbonLog
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

data class CarbonLog(val carbonMeasurements: List<CarbonMeasurement>): CoreCarbonLog<Instant> {
    fun copyAdding(measurement: CarbonMeasurement) = copyAddingMultiple(measurements = listOf(measurement))

    fun copyAddingMultiple(measurements: List<CarbonMeasurement>): CarbonLog =
        CarbonLog(carbonMeasurements = this.carbonMeasurements + measurements)

    override fun getRangeCarbonKgs(from: Instant, to: Instant, inclusive: Boolean): Double {
        val filter: (cm: CarbonMeasurement) -> Boolean = if (!inclusive) {
            { cm: CarbonMeasurement -> cm.dt > from && cm.dt < to }
        } else {
            { cm: CarbonMeasurement -> cm.dt in from..to }
        }

        val filtered = this.carbonMeasurements.filter(filter)
        return filtered.sumCO2Kgs()
    }

    override fun getTotalCarbonKgs(): Double = carbonMeasurements.sumCO2Kgs()

    override fun getCurrentYearCarbonKgs(): Double = carbonMeasurements
        .filter { cm: CarbonMeasurement -> cm.getYear() == LocalDate.now().year }
        .sumCO2Kgs()
}

private fun CarbonMeasurement.getYear(timezoneId: ZoneId = ZoneId.of("UTC")): Int =
    LocalDate.ofInstant(this.dt, timezoneId).year

private fun List<CarbonMeasurement>.sumCO2Kgs(): Double =
    when {
        this.isEmpty() -> 0.0
        else -> this.map { cm -> cm.co2Kg }
            .reduce { acc, next -> acc + next }
    }
